elliptic-curve/src/hash2field/expand_msg_xmd.rs

```diff
@@ -1,72 +1,97 @@
-use super::ExpandMsg;
+use super::{Domain, ExpandMsg};
 use core::marker::PhantomData;
-use digest::{
+use digest_traits::{
     generic_array::{typenum::Unsigned, GenericArray},
     BlockInput, Digest,
 };
-use subtle::{Choice, ConditionallySelectable};
 
 /// Placeholder type for implementing expand_message_xmd based on a hash function
-#[derive(Debug)]
-pub struct ExpandMsgXmd<HashT> {
+pub struct ExpandMsgXmd<HashT>
+where
+    HashT: Digest + BlockInput,
+{
+    b_0: GenericArray<u8, HashT::OutputSize>,
+    b_vals: GenericArray<u8, HashT::OutputSize>,
+    domain: Domain<HashT::OutputSize>,
+    index: usize,
+    offset: usize,
+    ell: usize,
     phantom: PhantomData<HashT>,
 }
 
+impl<HashT> ExpandMsgXmd<HashT>
+where
+    HashT: Digest + BlockInput,
+{
+    fn next(&mut self) -> bool {
+        if self.index < self.ell {
+            self.index += 1;
+            self.offset = 0;
+            // b_0 XOR b_(idx - 1)
+            let mut tmp = GenericArray::<u8, HashT::OutputSize>::default();
+            self.b_0
+                .iter()
+                .zip(&self.b_vals[..])
+                .enumerate()
+                .for_each(|(j, (b0val, bi1val))| tmp[j] = b0val ^ bi1val);
+            self.b_vals = HashT::new()
+                .chain(tmp)
+                .chain([self.index as u8])
+                .chain(self.domain.data())
+                .chain([self.domain.len() as u8])
+                .finalize();
+            true
+        } else {
+            false
+        }
+    }
+}
+
 /// ExpandMsgXmd implements expand_message_xmd for the ExpandMsg trait
-impl<HashT, const LEN_IN_BYTES: usize> ExpandMsg<LEN_IN_BYTES> for ExpandMsgXmd<HashT>
+impl<HashT> ExpandMsg for ExpandMsgXmd<HashT>
 where
     HashT: Digest + BlockInput,
 {
-    fn expand_message(msg: &[u8], dst: &[u8]) -> [u8; LEN_IN_BYTES] {
+    fn expand_message(msg: &[u8], dst: &'static [u8], len_in_bytes: usize) -> Self {
         let b_in_bytes = HashT::OutputSize::to_usize();
-        let ell = (LEN_IN_BYTES + b_in_bytes - 1) / b_in_bytes;
+        let ell = (len_in_bytes + b_in_bytes - 1) / b_in_bytes;
         if ell > 255 {
             panic!("ell was too big in expand_message_xmd");
         }
```

Comment:
I played around with how to circumvent this locally and I achieved some pretty ugly code to check this on a type level.

So I believe the goal is to keep these functions infallible, certainly not panicking. I don't know how else to do this without doing some ugly `where` bounds on the length. To do this we will need to keep the previous implementation you had of storing the `LEN_IN_BYTES` as a generic (probably using `ArrayLength` instead of const `usize`).

That way we can put all these kind of checks on the `ExpandMsgXmd` type itself and keep functions infallible.
Another check that's missing here:
https://www.ietf.org/archive/id/draft-irtf-cfrg-hash-to-curve-13.html#section-5.4.1-4
> [..] For correctness, H requires b <= s.

This could be easily checked on the type level: `HashT::OutputSize: IsLessOrEqual<HashT::BlockSize>`.

The constraint pointed above that is currently handled by panicking:
https://www.ietf.org/archive/id/draft-irtf-cfrg-hash-to-curve-13.html#section-5.4.1-6
> - len_in_bytes, the length of the requested output in bytes,
  not greater than the lesser of (255 * b_in_bytes) or 2^16-1.

Would be a bit harder, but not impossible.

Alternatively of course, we can just return a `Result`, but panicking is not a long-term solution, or am I missing something? Maybe I'm overcautious or not understanding something properly.

WDYT? I'm willing to help, either by making follow-up PR's or PR against your fork @mikelodder7.

Reply:
A follow PR would be great if we can somehow enforce it with type bounds. I'm not sure how to do this cleanly though. We could just change it to be a `Result` and declare it good enough. The odds of this happening are pretty low when using any hash function from RustCrypto. That won't stop someone from shooting themselves if they have a custom function that has a ridiculous OutputSize

Reply:
Maybe an easy solution would be just limit the `Outputsize` to something sane and be done with it. But yes, let's clear that in a follow-up.